### PR TITLE
[GEOS-9705] Allow GeoPackage WPS process to inline metadata and add OWS contexts for request and data

### PR DESCRIPTION
**Pull request by aaime**
See: https://osgeo-org.atlassian.net/browse/GEOS-9705

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Commits changing the REST API, or any configuration object, should check if the REST API docs (Swagger YAML files and classic documentation) need to be updated.

### CODE REVIEW

File: src/community/geopkg/src/main/resources/org/geoserver/geopkg/wps/owsContextRequest.ftl [hunk 1]
```diff
@@ -0,0 +1,65 @@
+{
+    "type":"FeatureCollection",
```
Comment (jodygarnett) [inline src/community/geopkg/src/main/resources/org/geoserver/geopkg/wps/owsContextRequest.ftl:2]: This is a very interesting pull requests, is this documenting the source of the geopackage?

File: src/community/geopkg/src/main/java/org/geoserver/geopkg/wps/OWSContextWriter.java [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+        this.annotations = gpkg.getExtension(SemanticAnnotationsExtension.class);
+        this.styleWorker = styleWorker;
+    }
+
+    /**
+     * Adds the WPS request context. Should be called first, as the per layer metadata references
+     * depend on it.
+     *
+     * @throws IOException
+     * @return
```
Comment (jodygarnett) [inline src/community/geopkg/src/main/java/org/geoserver/geopkg/wps/OWSContextWriter.java:121]: return not needed

File: src/community/geopkg/src/main/java/org/geoserver/geopkg/wps/OWSContextWriter.java [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+            annotations.addReference(ar);
+        } catch (SQLException e) {
+            throw new IOException(e);
+        }
+    }
+
+    private List<Map<String, Object>> getLayersModel(
+            LayerGroupInfo layerGroup, Map<String, LayerInfo> layerStyles) throws IOException {
+        List<Map<String, Object>> result = new ArrayList<>();
+        List<PublishedInfo> layers = layerGroup.getLayers();
```
Comment (jodygarnett) [inline src/community/geopkg/src/main/java/org/geoserver/geopkg/wps/OWSContextWriter.java:373]: thought: Are style drive groups going to cause confusion here? Thinking of an style that draws the same features multiple times to establish a z-order.

File: src/community/geopkg/src/main/resources/org/geoserver/geopkg/wps/owsContextRequest.ftl (hunk 1 above)
Comment (aaime) [reply src/community/geopkg/src/main/resources/org/geoserver/geopkg/wps/owsContextRequest.ftl:2]: It's a OWS context, in JSON format. IN particular, it's provenance, provides the original request that generated the GeoPackage.

File: src/community/geopkg/src/main/java/org/geoserver/geopkg/wps/OWSContextWriter.java (hunk 3 above)
Comment (aaime) [reply src/community/geopkg/src/main/java/org/geoserver/geopkg/wps/OWSContextWriter.java:373]: This is exactly what the OWS context in question needs to do, dump the right drawing order and associate it to styles. So a  dump of the layer groups structure is exactly that is needed. However, testing we found a small issue, in GeoServer the group is defined in painter's order, while OWS context unfortunately defined the opposite order, the first listed layer must be painted on top. Will have to make a small change for it.

File: src/community/geopkg/src/main/java/org/geoserver/geopkg/wps/OWSContextWriter.java (hunk 2 above)
Comment (aaime) [reply src/community/geopkg/src/main/java/org/geoserver/geopkg/wps/OWSContextWriter.java]: Fixed

File: src/community/geopkg/src/main/java/org/geoserver/geopkg/wps/OWSContextWriter.java (hunk 3 above)
Comment (aaime) [reply src/community/geopkg/src/main/java/org/geoserver/geopkg/wps/OWSContextWriter.java:371]: Order has been fixed now.

**Pull request merged**